src/i_oalsound.c

```diff
@@ -352,9 +352,242 @@ const char **I_OAL_GetResamplerStrings(void)
+        !alAuxiliaryEffectSloti ||
+        !alGenFilters ||
+        !alDeleteFilters ||
+        !alIsFilter ||
+        !alFilteri ||
+        !alFilterf)
+    {
+        return;
+    }
+
+    // Unload all effects first.
+
+    if (alIsAuxiliaryEffectSlot(uiEffectSlot))
+    {
+        for (int i = 0; i < MAX_CHANNELS; i++)
+        {
+            alSourcei(oal->sources[i], AL_DIRECT_FILTER, AL_FILTER_NULL);
+            alSource3i(oal->sources[i], AL_AUXILIARY_SEND_FILTER, AL_EFFECTSLOT_NULL, 0, AL_FILTER_NULL);
+        }
+
+        alDeleteAuxiliaryEffectSlots(1, &uiEffectSlot);
+    }
+
+    if (alIsEffect(uiEffect))
+    {
+        alDeleteEffects(1, &uiEffect);
+    }
+
+    if (alIsFilter(uiFilter))
+    {
+        alDeleteFilters(1, &uiFilter);
+    }
+
+    // Apply equalizer effect.
+
+    if (snd_equalizer != EQ_PRESET_OFF)
+    {
+        if (!alIsAuxiliaryEffectSlot(uiEffectSlot))
+        {
+            alGenAuxiliaryEffectSlots(1, &uiEffectSlot);
+        }
+
+        if (!alIsEffect(uiEffect))
+        {
+            alGenEffects(1, &uiEffect);
+        }
+
+        if (!alIsFilter(uiFilter))
+        {
+            alGenFilters(1, &uiFilter);
+        }
+
+        alEffecti(uiEffect, AL_EFFECT_TYPE, AL_EFFECT_EQUALIZER);
+
+        // AL_LOWPASS_GAIN actually controls overall gain for the filter it's
+        // applied to (OpenAL Effects Extension Guide 1.1, pp. 16-17, 135).
+
+        alFilteri(uiFilter, AL_FILTER_TYPE, AL_FILTER_LOWPASS);
+
+        // Gains vary from 0.126 up to 7.943, which means from -18dB attenuation
+        // up to +18dB amplification, i.e. 20*log10(gain).
+
+        #define DB_TO_GAIN(db) ((ALfloat)BETWEEN(0.126f, 7.943f, pow(10.f, db/20.f)))
+
+        alEffectf(uiEffect, AL_EQUALIZER_LOW_GAIN, DB_TO_GAIN(snd_eq_low_gain));
+        alEffectf(uiEffect, AL_EQUALIZER_MID1_GAIN, DB_TO_GAIN(snd_eq_mid1_gain));
+        alEffectf(uiEffect, AL_EQUALIZER_MID2_GAIN, DB_TO_GAIN(snd_eq_mid2_gain));
+        alEffectf(uiEffect, AL_EQUALIZER_HIGH_GAIN, DB_TO_GAIN(snd_eq_high_gain));
+
+        alEffectf(uiEffect, AL_EQUALIZER_LOW_CUTOFF, (ALfloat)snd_eq_low_cutoff);
+        alEffectf(uiEffect, AL_EQUALIZER_HIGH_CUTOFF, (ALfloat)snd_eq_high_cutoff);
+
+        alAuxiliaryEffectSloti(uiEffectSlot, AL_EFFECTSLOT_EFFECT, uiEffect);
+
+        for (int i = 0; i < MAX_CHANNELS; i++)
+        {
+            // Mute the dry path.
+            alFilterf(uiFilter, AL_LOWPASS_GAIN, 0.0f);
+            alSourcei(oal->sources[i], AL_DIRECT_FILTER, uiFilter);
+
+            // Keep the wet path.
+            alFilterf(uiFilter, AL_LOWPASS_GAIN, 1.0f);
+            alSource3i(oal->sources[i], AL_AUXILIARY_SEND_FILTER, uiEffectSlot, 0, uiFilter);
+        }
+    }
+}
+
+void I_OAL_EqualizerPreset(void)
+{
+    struct
+    {
+        int *var;
+        int val[NUM_EQ_PRESETS];
+    } eq_presets[] =
+    {                        // Off, Bass, Bass+Treb, Mid
+        {&snd_eq_low_gain,    {   0,    2,    2,    0}},
```

Comment:
Guys, please lend me an ear, literally. 😁 Technically, I consider this PR as ready, but the presets need some more fine-tuning. However, my hearing is not good enough to act as a reference. Please tell me your prefered settings (and names) so we can merge this with a reasonable set of presets. Thank you! @ak-988 @ceski-1 @rfomin @MrAlaux

Reply:
The effect is too subtle for me, I would increase parameters x2 in the presets. I don't have good hearing either.

Reply:
Yeah, as I mentioned above, 1-3 dB changes are barely perceptible so it makes sense the current presets don't seem to do much. I want to plot the default EE EQ curve (it's a different 3-band method based on public domain code) and see what it would take to match that using OpenAL values. Then, if it makes sense, it could become a preset, probably something like "Bass boost, Treble cut".

Reply:
Also, this may be out of scope for this PR, but maybe it would be nice to have an EQ option for music if FluidSynth, OPL3, or other file formats (mp3, etc.) are used. We have no control over native MIDI, that's up to the OS.

Reply:
Here's a preset that's roughly similar to Eternity's default EQ:

```
snd_eq_low_gain     2
snd_eq_mid1_gain    0
snd_eq_mid2_gain    0
snd_eq_high_gain    -1
snd_eq_low_cutoff   340
snd_eq_high_cutoff  10000
```
Neither one sounds very different from a flat EQ to me. Just a slight bass boost that's more noticeable with a pair of speakers + subwoofer.

I think it's close enough when playing white noise:

![woof_eq](https://github.com/fabiangreffrath/woof/assets/56656010/22d3e186-9406-410a-9599-d90be3b2e788)

Reply:
The above plot was with 22kHz white noise. This one is with 44.1kHz white noise:

![woof_eq_44 1k](https://github.com/fabiangreffrath/woof/assets/56656010/0df984bd-e9b2-4653-8144-81ba68dfcb63)

Not sure why EE has that peak after 10kHz.

Reply:
Thank you very much for your efforts @ceski-1, but maybe it makes more sense to find our own generic equalizer presets than trying to run behind EE.

Reply:
Actually, I agree. The difference is subtle to me with the default sounds. I wanted to put an end to those comparisons.

Reply:
I exposed the remaining settings since they're useful for making proper adjustments in the mid range. I also increased the gain for the temporary presets just so we can get a feel for them. I hope you don't mind.

Here's a useful diagram from a manual [mentioned](https://github.com/kcat/openal-soft/blob/41e7f27337e49f72da5b2af575dca8bc3b6a23cf/alc/effects/equalizer.cpp#L50) in the OpenAL Soft source code:

![proteus_vx_manual](https://github.com/fabiangreffrath/woof/assets/56656010/fe428025-172a-4bcc-8465-6cddcb8c6189)

Reply:
> I exposed the remaining settings since they're useful for making proper adjustments in the mid range. I also increased the gain for the temporary presets just so we can get a feel for them. I hope you don't mind.

Of course not, that's highly appreciated! 👍

Reply:
> Guys, please lend me an ear, literally.

Just noticed you mentioned me. I'm not much of an audio expert, if at all; most I've done is play with the equalizer and a preinstalled sound tuner of sorts on my machine, for listening to music (although it probably affects all system audio). What could I do here?

Reply:
I think I'll make a placeholder menu. This isn't easy without some sliders that can be adjusted in real-time.

Reply:
I added a placeholder menu here: https://github.com/fabiangreffrath/woof/pull/1714/commits/996e713a44bc21269ccb8c2f0cf1fe5ab4a6362d Select an equalizer preset from General > Audio. Then fine tune with the sliders in General > EQ. Please be careful with your ears.

Reply:
> What could I do here?

Just play around with the equalizer settings and presets. Thankfully, @ceski-1 has taken the lead and exposed the latter in the menu. The goal is to find some 3 or 4 default equalizer presets that everybody could agree upon.

Reply:
From testing so far, I like the "bass+treble" preset for linear sampling the most. It has a bit more weight in the shotgun for instance, and a hint of pleasant crunch reminiscent of the Sound Blaster 🍷. "Bass" actually makes nearest sampling not too bad. There's a ton of possible combinations now, so lots to potentially try. Maybe "mid" doesn't have as much of a niche, I might try a B+T variation with different frequencies but am unsure how noticeable it would be.

Reply:
Please feel free to improve the presets and suggest your changes here.

Reply:
https://github.com/fabiangreffrath/woof/pull/1714/commits/6916de4f4ad8c553359eac989d8005f37d4c528a Here's an alternate set that mimics "classical", "rock", and "vocal" presets typically seen with sound cards and music players. It's roughly similar to the existing "bass", "bass+treble", and "mid" presets. I set every pre-amp to zero so it's easier to make quick comparisons.

I'm not sure how much more adjustment is needed, maybe one more round of feedback and I think that's good enough. If there ends up being a sub-menu for EQ, I think the bandwidth settings can be kept as config-only.

Reply:
> Here's an alternate set that mimics "classical", "rock", and "vocal" presets typically seen with sound cards and music players. 

Thank you very much! I wouldn't mind if we called the presets like this. It's a bit less abstract.

> I'm not sure how much more adjustment is needed, maybe one more round of feedback and I think that's good enough. If there ends up being a sub-menu for EQ, I think the bandwidth settings can be kept as config-only.

Agreed.

Reply:
I like the updated presets. Two general things I've realized: 1) what sounds cool in isolation isn't necessarily good for extended periods of playing, and 2) the "ideal" EQ really seems to depend on what you're listening with, my cheap Bluetooth headphones have their own bass boost apparently and Woof's extra is too much, but my nice neutral Sennheisers benefit from added EQ. So, I make some small tweaks to taste but nothing I'd suggest as an "improvement" over what's already here.

It's so nice having some thunder back in the chaingun and really feeling that resonant, blubbery cacodemon death. For me the sound just needed that extra nudge, and having the settings exposed one way or another means anyone who wants to figure it out can get what they want.

Reply:
> I set every pre-amp to zero so it's easier to make quick comparisons.

What would these values be realistically?

Reply:
I'd have to calculate it again but I think it was roughly 0, -4, -5, -4 dB from left to right.

Reply:
Alright. I don't think we will get any more feedback for this PR, so what do you say, how should we proceed?

Reply:
If this remains as a preset only, there's not much left to do. But if we want to expose the EQ settings, I think it needs a sub menu with a vertical slider layout, otherwise it gets confusing very quickly. That might be a lot of work for a small feature. So maybe we need to experiment with the sub menu idea some more and see if it makes sense to continue.

Reply:
So, what's left to do for this PR:

- Give EQ presets some less abstract names
- Adjust pre-amp values for EQ presets
- Comment out or remove EQ menu
- Add EQ preset selection to General/Audio menu

Reply:
> I think it needs a sub menu with a vertical slider layout

Do we really need a vertical layout? If horizontal is good enough, then it's not that hard: #1743 I'm busy these days and haven't done much here.

Reply:
> Do we really need a vertical layout? If horizontal is good enough, then it's not that hard: #1743 I'm busy these days and haven't done much here.

I'd like to get this PR through the door with just the EQ presets selection in the menu first.

Reply:
I'd consider this done now, ready for review.

Reply:
> I'd consider this done now, ready for review.

I made a few small changes, but it looks good to me.

Reply:
Thank you, I love how you cleaned this up!